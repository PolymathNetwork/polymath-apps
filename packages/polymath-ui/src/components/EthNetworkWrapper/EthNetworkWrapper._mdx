---
name: EthNetworkWrapper
---

import { Playground, PropsTable } from 'docz';
import EthNetworkWrapper from './';

# Footer

<PropsTable of={EthNetworkWrapper} />

## Basic usage

<Playground>
  <EthNetworkWrapper />
</Playground>
