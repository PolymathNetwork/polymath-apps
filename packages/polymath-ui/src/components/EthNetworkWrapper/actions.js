// @flow

import Web3 from 'web3';

import getNetwork, { NETWORK_LOCAL, NETWORK_LOCALVM } from './networks';
import type { ExtractReturn } from './helpers';
import Contract, { setupContracts } from '@polymathnetwork/js';
import { txHash, txEnd } from '../TxModal/actions';

export type NetworkParams = {|
  id: number,
  name: string,
  account: string,
  web3: Web3,
  web3WS: Web3,
|};

export const CONNECTED = 'polymath-auth/CONNECTED';
const connected = (params: NetworkParams) => ({ type: CONNECTED, params });

export const FAIL = 'polymath-auth/FAIL';
const fail = (error: number) => ({ type: FAIL, error });

export type Action =
  | ExtractReturn<typeof connected>
  | ExtractReturn<typeof fail>;

const web3 = new Web3();
const web3WS = new Web3(); // since MetaMask doesn't support WebSockets we need this extra client for events subscribing

export const ERROR_NOT_INSTALLED = 1;
export const ERROR_LOCKED = 2;
export const ERROR_NETWORK = 3;
export const ERROR_DISCONNECTED = 4;

const initPolymathJs = async (params: {
  networkParams: Object,
  polymathRegistryAddress: string,
  dispatch: Function,
}) => {
  const { networkParams, polymathRegistryAddress, dispatch } = params;
  Contract.setParams({
    ...networkParams,
    txHashCallback: hash => dispatch(txHash(hash)),
    txEndCallback: receipt => dispatch(txEnd(receipt)),
  });

  // initialize polymath-js contracts with addresses from the polymath registry
  await setupContracts(polymathRegistryAddress);
};

export const init = (networks: Array<string>) => async (dispatch: Function) => {
  try {
    let id = undefined;
    if (typeof window.web3 !== 'undefined') {
      // Metamask/Mist
      web3.setProvider(window.web3.currentProvider);
      id = await web3.eth.net.getId();
    }

    const isLocalhost =
      String(id) === NETWORK_LOCAL ||
      String(id) === NETWORK_LOCALVM ||
      id === undefined;

    const network = getNetwork(id);

    if (
      network === undefined ||
      (!isLocalhost && !networks.includes(String(id)))
    ) {
      throw new Error(ERROR_NETWORK);
    }
    web3WS.setProvider(process.env.REACT_APP_NODE_WS || network.url);
    if (!id) {
      web3.setProvider(web3WS.currentProvider);
      id = await web3.eth.net.getId();
    }

    const name = network.name;
    const [account] = await web3.eth.getAccounts();
    if (id) {
      // https://github.com/MetaMask/faq/blob/master/DEVELOPERS.md#ear-listening-for-selected-account-changes
      setInterval(() => {
        web3.eth.getAccounts().then(accounts => {
          if (accounts[0] !== account) {
            window.location.reload();
          }
        });
        web3.eth.net.getId().then(_id => {
          if (id !== _id) {
            window.location.reload();
          }
        });
      }, 100);
    }
    if (!account) {
      throw new Error(ERROR_LOCKED);
    }

    // TODO @bshevchenko: https://github.com/INFURA/infura/issues/80 hack below
    web3WS.eth.subscribe('newBlockHeaders', error => {
      if (error) {
        // eslint-disable-next-line
        console.error('web3WS newBlockHeaders', error);
        dispatch(fail(ERROR_DISCONNECTED));
      }
    });

    const networkParams = {
      id,
      name,
      account,
      web3,
      web3WS,
    };

    await initPolymathJs({
      networkParams,
      dispatch,
      polymathRegistryAddress: network.polymathRegistryAddress,
    });

    dispatch(connected(networkParams));
  } catch (e) {
    if (![ERROR_LOCKED, ERROR_NETWORK].includes(Number(e.message))) {
      // eslint-disable-next-line
      console.error('polymath-auth', e);
      e.message = ERROR_NOT_INSTALLED;
    }
    dispatch(fail(e.message));
  }
};
