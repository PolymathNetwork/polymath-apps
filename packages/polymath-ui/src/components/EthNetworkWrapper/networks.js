// @flow

type Network = {
  name: string,
  url: string,
};

export const NETWORK_MAIN = '1';
export const NETWORK_ROPSTEN = '3';
export const NETWORK_RINKEBY = '4';
export const NETWORK_KOVAN = '42';

export function getNetworkInfos(id: string = 'local'): Network {
  const networks = {
    [NETWORK_MAIN]: {
      name: 'Mainnet',
      url: process.env.REACT_APP_NETWORK_MAIN_WS,
    },
    [NETWORK_ROPSTEN]: {
      name: 'Ropsten Testnet',
      url: 'wss://ropsten.infura.io/ws',
    },
    [NETWORK_RINKEBY]: {
      name: 'Rinkeby Testnet',
      url: 'wss://rinkeby.infura.io/ws',
    },
    [NETWORK_KOVAN]: {
      name: 'Kovan Testnet',
      url: process.env.REACT_APP_NETWORK_KOVAN_WS,
    },
    local: {
      name: 'Localhost',
      url: process.env.REACT_APP_NETWORK_LOCAL_WS,
    },
  };

  return networks[id];
}
