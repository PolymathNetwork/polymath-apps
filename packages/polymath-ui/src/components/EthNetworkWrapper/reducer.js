// @flow

import Web3 from 'web3';

import type { Action } from './actions';
import { CONNECTED, FAILED } from './';

export type NetworkState = {|
  isConnected: boolean,
  isFailed: boolean,
  error: ?number,
  id: ?number,
  name: ?string,
  account: ?string,
  web3: ?Web3,
  web3WS: ?Web3,
|};

const defaultState: NetworkState = {
  isConnected: false,
  isFailed: false,
  error: null,
  id: null,
  name: null,
  account: null,
  web3: null,
  web3WS: null,
};

export default (state: NetworkState = defaultState, action: Action) => {
  switch (action.type) {
    case CONNECTED:
      return {
        ...state,
        isConnected: true,
        isFailed: false,
        ...action.params,
      };
    case FAILED:
      return {
        ...state,
        isConnected: false,
        isFailed: true,
        error: action.error,
      };
    default:
      return state;
  }
};
