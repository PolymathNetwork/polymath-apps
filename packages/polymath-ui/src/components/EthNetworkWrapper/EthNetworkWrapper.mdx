---
name: EthNetworkWrapper
---

import { Playground, PropsTable } from 'docz';
import EthNetworkWrapper from './';

# Footer

<PropsTable of={Footer} />

## Basic usage

<Playground>
  <EthNetworkWrapper />
</Playground>
