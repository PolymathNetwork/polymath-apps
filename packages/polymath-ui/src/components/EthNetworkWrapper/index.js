// @flow

import EthNetworkWrapper from './EthNetworkWrapper';

export { default as reducer } from './reducer';

export const ERROR_NOT_INSTALLED = 1;
export const ERROR_LOCKED = 2;
export const ERROR_NETWORK = 3;
export const ERROR_DISCONNECTED = 4;
export const ERROR_ACCESS_REQUESTED = 5;

export {
  NETWORK_MAIN,
  NETWORK_KOVAN,
  NETWORK_RINKEBY,
  NETWORK_ROPSTEN,
} from './networks';

export default EthNetworkWrapper;

export type { NetworkState } from './reducer';
export type { NetworkParams } from './actions';
