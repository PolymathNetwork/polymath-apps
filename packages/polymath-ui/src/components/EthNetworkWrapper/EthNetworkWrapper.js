// @flow

// eslint-disable-next-line
import React, { Component } from 'react';
import { connect } from 'react-redux';
import canUseDOM from 'can-use-dom';

import type { Node } from 'react';

import { init } from './actions';

type StateProps = {|
  isConnected: boolean,
  isFailed: boolean,
  error: ?number,
|};

type DispatchProps = {|
  init: (networks: Array<string>) => any,
|};

const mapStateToProps = (state): StateProps => ({
  isConnected: state.network.isConnected,
  isFailed: state.network.isFailed,
  error: state.network.error,
});

const mapDispatchToProps: DispatchProps = {
  init,
};

type Props = {|
  children: Node,
  loading: Node,
  guide: Node,
  networks: Array<string>,
|} & StateProps &
  DispatchProps;

class EthNetworkWrapper extends Component<Props> {
  init = () => {
    this.props.init(this.props.networks);
  };

  componentWillMount() {
    if (!this.props.isConnected && !this.props.isFailed) {
      if (canUseDOM) {
        if (document.readyState === 'complete') {
          this.init();
        } else {
          window.addEventListener('load', () => {
            this.init();
          });
        }
      }
    }
  }

  render() {
    if (this.props.isConnected) {
      return this.props.children;
    }
    if (!this.props.isFailed) {
      return this.props.loading;
    }
    return this.props.guide;
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EthNetworkWrapper);
