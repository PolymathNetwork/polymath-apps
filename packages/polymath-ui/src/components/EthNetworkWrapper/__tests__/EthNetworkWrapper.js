import React from 'react';
import renderer from 'react-test-renderer';

import { EthNetworkWrapper } from '../EthNetworkWrapper';

// NOTE @RafaelVidaurre: Muted this test as it is tackled in the 2.0 branch
describe('EthNetworkWrapper', () => {
  test.skip('renders without crashing', () => {
    const loadingContent = <div>Loading...</div>;
    const component = renderer.create(
      <EthNetworkWrapper loading={loadingContent} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
