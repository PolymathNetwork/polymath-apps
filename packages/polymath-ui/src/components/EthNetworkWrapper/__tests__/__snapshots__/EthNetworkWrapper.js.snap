// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthNetworkWrapper renders without crashing 1`] = `
<div>
  Loading...
</div>
`;
